src/Apps/Legion/Components/CookSource/CookSource.tsx

```diff
@@ -0,0 +1,284 @@
+import { useTranslation } from 'react-i18next';
+import {
+    PIDSourceOptions,
+    SourceType,
+    SourceTypeOptions,
+    TIMESTAMP_COLUMN_NAME
+} from '../DataPusher/DataPusher.types';
+import ClusterPicker from '../Pickers/ClusterPicker/ClusterPicker';
+import DatabasePicker from '../Pickers/DatabasePicker/DatabasePicker';
+import TablePicker from '../Pickers/TablePicker/TablePicker';
+import useAdapter from '../../../../Models/Hooks/useAdapter';
+import {
+    IGetTableAdapterParams,
+    ITable
+} from '../../Adapters/Standalone/DataManagement/Models/DataManagementAdapter.types';
+import { useADXAdapter } from '../../Hooks/useADXAdapter';
+import {
+    IADXAdapterTargetContext,
+    IADXConnection,
+    ICookSource,
+    IPIDDocument
+} from '../../Models';
+import produce from 'immer';
+import { WritableDraft } from 'immer/dist/internal';
+import { getTableSchemaTypeFromTable } from '../../Services/DataPusherUtils';
+import { AppDataContext } from '../../Contexts/AppDataContext/AppDataContext';
+
+const debugLogging = false;
+const logDebugConsole = getDebugLogger('CookSource', debugLogging);
+
+const defaultTimeSeriesSource: ICookSource = {
+    cluster: null,
+    database: null,
+    table: null,
+    twinIdColumn: null,
+    tableType: null
+};
+
+const defaultPIDSource: ICookSource = {
+    pidUrl: ''
+};
+
+const CookSource: React.FC<ICookSourceProps> = (props) => {
+    const {
+        onSourceTypeChange,
+        onSourceChange,
+        onGetTableData,
+        targetAdapterContext = AppDataContext,
+        isClusterVisible = true
+    } = props;
+
+    // state
+    const [selectedSourceType, setSelectedSourceType] = useState<SourceType>(
+        SourceType.Timeseries
+    );
+    const [selectedSource, setSelectedSource] = useState<ICookSource>(
+        defaultTimeSeriesSource
+    );
+    const [sourceTableColumnOptions, setSourceTableColumnOptions] = useState<
+        Array<IDropdownOption>
+    >([]);
+
+    // hooks
+    const { t } = useTranslation();
+    const adapter = useADXAdapter(
+        targetAdapterContext as React.Context<IADXAdapterTargetContext>
+    );
+
+    const getTableState = useAdapter({
+        adapterMethod: (params: IGetTableAdapterParams) =>
+            adapter.getTable(
+                params.databaseName,
+                params.tableName,
+                TIMESTAMP_COLUMN_NAME
+            ),
+        isAdapterCalledOnMount: false,
+        refetchDependencies: [adapter]
+    });
+
+    // callbacks
+    const handleSourceTypeChange = useCallback(
+        (_event, option: IDropdownOption) => {
+            setSelectedSourceType(option.key as SourceType);
+            onSourceTypeChange?.(option.key as SourceType);
+            const newSource =
+                option.key === SourceType.Timeseries
+                    ? defaultTimeSeriesSource
+                    : defaultPIDSource;
+            setSelectedSource(newSource);
```

Comment:
should we create a wrapper to do the "SetSource" operation so it will always call both the local state set and the notification of the callback? I see we repeat that same line in several places, is it worth abstracting so we don't ever set the value wihtout notifying?

Reply:
added use-effects for it